### PR TITLE
[DO NOT SQUASH] Import MHAL - MLIR Hardware Abstraction Layer

### PR DESCRIPTION
**Pull request by sjw36**
XModel has been extracted out as a separate repo: mlir-hal [https://github.com/ROCmSoftwarePlatform/mlir-hal]

This PR removes XModel and re-aligns with upstream LLVM/MLIR.

Still some upstream changes will be needed to support mhal.launch through bufferization.

### CODE REVIEW

**Comment by sjw36** [comment]
Implements https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/890

**Comment by sjw36** [comment]
I think this is going to pass now. Please commence reviews. @krzysz00 @jerryyin @giuseros

File: cmake/llvm-project.cmake [hunk 1]
```diff
@@ -61,6 +61,7 @@ list(APPEND LLVM_INCLUDE_DIRS
 
 # Linker flags
 set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-rpath -Wl,${LLVM_BINARY_DIR}/llvm/lib")
+list(APPEND CMAKE_BUILD_RPATH "${LLVM_BINARY_DIR}/llvm/lib")
```
Comment (krzysz00) [inline cmake/llvm-project.cmake:63]: Can we double check the LLVM files to see what impact this has on our build?

File: external/llvm-project/mlir.diff
Comment (krzysz00) [inline external/llvm-project/mlir.diff:1]: Delete pointless file

File: external/mlir-hal/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+set(MHAL_MAIN_INCLUDE_DIR "${MHAL_PROJECT_DIR}/include")
+option(MHAL_ENABLE_TRANSFORMS "Enable MHAL transforms" ON)
+
+set(LLVM_ENABLE_RTTI OFF CACHE BOOL "")
+
+include_directories("${MHAL_MAIN_INCLUDE_DIR}")
+include_directories("${MHAL_BINARY_DIR}/include")
+
+add_subdirectory(include/mlir)
+add_subdirectory(lib)
```
Comment (krzysz00) [inline external/mlir-hal/CMakeLists.txt:17]: Note to future us about making this buildable standalone. Can you set up a ticket for that?

File: external/mlir-hal/include/mlir/Conversion/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+
+set(LLVM_TARGET_DEFINITIONS MHALPasses.td)
+mlir_tablegen(MHALPasses.h.inc -gen-pass-decls -name MHALConversion)
+#mlir_tablegen(MHALPasses.capi.h.inc -gen-pass-capi-header --prefix MHALConversion)
```
Comment (krzysz00) [inline external/mlir-hal/include/mlir/Conversion/CMakeLists.txt:4]: Let's go ahead and enable a C API?

File: external/mlir-hal/include/mlir/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+add_subdirectory(Conversion)
```
Comment (krzysz00) [inline external/mlir-hal/include/mlir/CMakeLists.txt:1]: I'm going to go ahead and hang "Why not `include/mhal/`?" off of this.

**Review by krzysz00 — APPROVED** [review]
I'm not seeing any obvious issues with the refactor, though I still don't quite like the design of `mhal`'s packaging format and would really like to move us to a "one module per target, which includes the relevant host code" system, but that's outside the scope of the PR.

File: cmake/llvm-project.cmake (hunk 1 above)
Comment (sjw36) [reply cmake/llvm-project.cmake:63]: This setting is the proper way to encode the RPATHs into the binaries for BUILD. I have removed the previous line as it is no longer necessary.

We may need to set CMAKE_INSTALL_RPATH for installation runtime paths.

**Comment by sjw36** [comment]
See ticket for mlir-hal project improvements: https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/983

**Pull request merged**